### PR TITLE
Logsumexp fixes - allows sign flags and b==0

### LINKED ISSUE #4702: logsumexp with zero scaling factor
```python
>>> scipy.misc.logsumexp([-1000])
-1000.0
>>> scipy.misc.logsumexp([-1000, 0], b=[1, 0])
-inf
```

### PR DESCRIPTION
**Pull request by aarchiba**
Fixes #4702 and #4627

### CODE REVIEW

**Comment by argriffing** [comment]
Great! If this passes its tests then I agree that it should close https://github.com/scipy/scipy/issues/4702.  One faux-pas in this PR is that some tests use raw `assert` instead of `numpy.testing.assert_` or other assertions from numpy.testing. I don't remember why that's important.

**Comment by charris** [comment]
The python `assert

**Comment by aarchiba** [comment]
Sorry. Astropy uses py.test, which is smart enough about introspection to make assert a==b do the right thing, even so far as reporting the values if they differ.

**Comment by rgommers** [comment]
Not sure what's the default for Python binaries on Windows, but there is at least one popular web tool (forgot which one) that runs the Scipy stack and uses ``-OO``. I thought that the statements get stripped out of bytecode; py.test cannot work in that case either right?

    $ python -c "assert 1 == 2"
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    AssertionError
    $ python -O -c "assert 1 == 2"
    <no output>

File: scipy/misc/common.py [hunk 1]
```diff
@@ -79,9 +89,20 @@ def logsumexp(a, axis=None, b=None, keepdims=False):
     9.9170178533034665
     >>> np.log(np.sum(b*np.exp(a)))
     9.9170178533034647
+
+    Returning a sign flag
+
+    >>> logsumexp([1,2],b=[1,-1])
+    (1.5413248546129181, -1)
```
Comment (ev-br) [inline scipy/misc/common.py]: Seems inconsistent with return_sign defaulting to False.

File: scipy/misc/common.py [hunk 2]
```diff
@@ -41,17 +41,27 @@ def logsumexp(a, axis=None, b=None, keepdims=False):
...
         Scaling factor for exp(`a`) must be of the same shape as `a` or
-        broadcastable to `a`.
+        broadcastable to `a`. These values may be negative in order to
+        implement subtraction.
 
         .. versionadded:: 0.12.0
+    return_sign : bool, optional
+        If this is set to True, the result will be a pair containing sign
+        information; if False, results that are negative will be returned
+        as NaN.
```
Comment (ev-br) [inline scipy/misc/common.py]: Can you mention the default here pleased

**Comment by aarchiba** [comment]
Doctests are definitely not getting run. Any idea why?

**Comment by pv** [comment]
Scipy doesn't use doctests for testing, it's all meant only for
documentation.

**Comment by aarchiba** [comment]
@pv , doesn't that mean that examples in the docs get out of sync with the code?

**Comment by charris** [comment]
@aarchiba Yes. But it is considered a lesser evil. The nosetests are clearer, easier to maintain,  easier to keep compatible for both Python3 and Python2, and don't depend on such things as how many digits get printed. That said, I think it is possible to run the doctests so that they can be checked for correctness, but we don't depend on them for testing.

**Comment by aarchiba** [comment]
@pv Indeed, doctests are not so useful as tests of the code, but they do mean that continuous integration checks that (some) documentation matches functionality. refguide_check.py is (now) able to check at least the syntax of documentation examples, but this was implemented because many of them were broken. I guess at this point making scipy pass doctests would be a huge project.

**Comment by pv** [comment]
I think you can do `DISPLAY= python runtests.py --doctests -s linalg`
and see what comes out. It's not very convenient though (e.g. doesn't
switch matplotlib backend so many figures pop up), as I this is not
something that has been regularly used. Probably fixable though if
someone wants to take a shot at it.
.
In principle, it could be added to travis. But you are right that there
are currently quite many failures. Probably takes a bit of time to fix
that, but still probably doable.
.
The tradeoff here is that one should not clutter the documentation too
much with the doctests extra markup. Pseudocode is probably also
acceptable for some purposes. There are several options in doctest that
could be turned on by default, to avoid too much mess, but I don't think
they are currently on by default.

**Comment by ev-br** [comment]
Just for the fun of it, I tried using `assert_allclose` for doctesting. With https://github.com/ev-br/scipy/tree/doctests_on_steroids, `$ python tools/refguide_check.py --check_docs linalg` produces https://gist.github.com/ev-br/0fb985cb3f2fb1f26d16.

Something fishy is going on in the new QR updating routines :-). Other than that, there are a couple of examples which are off; tightening up the comparison tolerance produces *a lot* of sorta-failures.

**Comment by ev-br** [comment]
Doctests aside, this fixes two annoying limitations of `logsumexp` so in it goes.  Thanks @aarchiba, all.

**Comment by ev-br** [comment]
Actually, these fixes should probably be mentioned in the release notes for scipy 0.17.

**Pull request merged**